### ISSUE TITLE
Generic protocol is incorrectly shown as incompatible under certain cases

### ISSUE THREAD

**Issue by rshin** [microsoft/pyright#3444]
**Describe the bug**
A class which should be marked as implementing a generic protocol is not identified as such.

**To Reproduce**
```python
from typing import Generic, TypeVar, Union, Protocol

T = TypeVar("T")
U = TypeVar("U")

class X(Generic[T, U], Protocol):
    def f(self) -> Union[T, U]:
        ...

    def g(self) -> "X[T, U]":
        ...

class Y(Generic[T, U]):
    def f(self) -> Union[T, U]:
        raise NotImplementedError

    def g(self) -> X[T, U]:
        return Y[T, U]()
```

results in the following: 
```
error: Expression of type "Y[U@Y, T@Y]" cannot be assigned to return type "X[T@Y, U@Y]"
    "Y[U@Y, T@Y]" is incompatible with protocol "X[T@Y, U@Y]"
      TypeVar "T@X" is invariant
        Type "U@Y" cannot be assigned to type "T@Y" (reportGeneralTypeIssues)
```

The following variation results in a different error:
```python
from dataclasses import dataclass
from typing import Generic, TypeVar, Union, Protocol

T = TypeVar("T")
U = TypeVar("U")

class X(Generic[T, U], Protocol):
    def f(self) -> Union[T, U]:
        ...

    def g(self, tu: Union[T, U]) -> "X[T, U]":
        ...

@dataclass
class Y(Generic[T, U]):
    tu: Union[T, U]
    def f(self) -> Union[T, U]:
        raise NotImplementedError

    def g(self, tu: Union[T, U]) -> X[T, U]:
        return Y[T, U](tu)
```

The error is
```
error: Expression of type "Y[T@Y, U@Y]" cannot be assigned to return type "X[T@Y, U@Y]"
    "Y[T@Y, U@Y]" is incompatible with protocol "X[T@Y, U@Y]"
      "g" is an incompatible type
        Type "(tu: T@Y | U@Y) -> X[T@Y, U@Y]" cannot be assigned to type "(tu: T@X | U@X) -> X[T@X, U@X]"
          Function return type "X[T@Y, U@Y]" is incompatible with type "X[T@X, U@X]"
            TypeVar "T@X" is invariant (reportGeneralTypeIssues)
```

Replacing `Union` with `Dict`, removing `f`, or changing the return type of `f` to something else like `T` or `U` all cause the error to go away.

**VS Code extension or command-line**
pyright 1.1.245

**Comment by erictraut** [comment]
Thanks for the bug report.

The issue here is that there's an ambiguity when performing type compatibility checks for a union that contains TypeVars, such as `T | U`. Consider, for example, if you attempt to assign type `int | str` to type `T | U`. There are four possible solutions in this case. 
1. `T` is `int` and `U` is `str`
2. `T` is `str` and `U` is `int`
3. `T` is `int | str` and `U` is unsolved
4. `T` is unsolved and `U` is `int | str`

Pyright incorporates a number of heuristics that attempt to pick the most natural solution in ambiguous cases like this. I've modified these heuristics so the above code samples type check as expected.

This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.247, which I just published. It will also be included in the next release of pylance.
